### PR TITLE
Username was not unique

### PR DESCRIPTION
**Pull request by sidgoyal**
Looking at the Authentication Collection, it turns out that the usernames were not unique. Created a fix to have unique usernames, and also added a test for it. removed the findByUsernameAndPassword method in AuthenticationRepository because its redundant now with findByUsername and unique usernames

### CODE REVIEW

**Comment by amitmawkin** [comment]
They are unique if you try to add a user that already exist you will get following error message 
screen shot 2015-10-05 at 6 30 08 pm

can you please check what version of code you are using ..

thanks,

**Comment by sidgoyal** [comment]
I just recreated the error from the UI 

image


can you try running the dashboard with a new fresh mongoDB instance ? also, you can pull the test that I added and try running it, without the change in the model, it should fail.

File: common/src/test/java/com/capitalone/dashboard/repository/AuthenticationRepositoryTest.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    	try{
+    	Authentication user2 = new Authentication(username, "pass2");
+    	
+    	// This line should throw a DuplicateKeyException
+    	authenticationRepository.save(user2);
+    	
+    	// If the above line did not throw a DuplicateKeyException, fail the test
+    	fail("Didn't throw any Exception");
+    	}
+    	catch(DuplicateKeyException e){
```
Comment (jdamick) [inline common/src/test/java/com/capitalone/dashboard/repository/AuthenticationRepositoryTest.java]: remove this catch, use: http://junit.sourceforge.net/javadoc/org/junit/Test.html#expected()

File: api/src/main/java/com/capitalone/dashboard/service/AuthenticationServiceImpl.java [hunk 2]
```diff
@@ -72,11 +70,9 @@ public void delete(String username) {
     @Override
     public boolean authenticate(String username, String password) {
         boolean flag = false;
-        List<Authentication> authenticationList = authenticationRepository.findByUsernameAndPassword(username, password);
-        if (authenticationList.size() == 1) {
-            if (authenticationList.get(0).getUsername().equals(username) && authenticationList.get(0).getPassword().equals(password)) {
-                flag = true;
-            }
+        Authentication authentication = authenticationRepository.findByUsername(username);
+        if (authentication.getUsername().equals(username) && authentication.getPassword().equals(password)) {
```
Comment (jdamick) [inline api/src/main/java/com/capitalone/dashboard/service/AuthenticationServiceImpl.java]: if you just looked it up by username why do you need to .equals it?

Comment (jdamick) [inline api/src/main/java/com/capitalone/dashboard/service/AuthenticationServiceImpl.java]: can authentication be null?

**Comment by jdamick** [comment]
either way, this is a good test that we should have.  thank you for identifying these issues.  it looks like we are comparing the clear-text password in here? @amitmawkin ?  that would be a major TODO for us...

**Comment by amitmawkin** [comment]
I will investigate as I cannot produce this on my end

File: common/src/main/java/com/capitalone/dashboard/model/Authentication.java [hunk 3]
```diff
@@ -11,8 +12,9 @@
 public class Authentication extends BaseModel {
 	
 	
-
+	@Indexed(unique = true)
```
Comment (jdamick) [inline common/src/main/java/com/capitalone/dashboard/model/Authentication.java]: good catch

**Comment by amitmawkin** [comment]
I checked on my end, i have an index turned on authentication collection.
so probably that explain why i am not facing this on my local mongodb collection
{
    "username" : 1,
    "password" : 1
}

File: common/src/test/java/com/capitalone/dashboard/repository/AuthenticationRepositoryTest.java (hunk 1 above)
Comment (sidgoyal) [inline common/src/test/java/com/capitalone/dashboard/repository/AuthenticationRepositoryTest.java]: Yeah, @Expected does not tell which line threw the exception, that's why I used the catch explicitly for the second save() call.

**Comment by sidgoyal** [comment]
@jdamick  So I made the changes that you suggested. Also did some refactoring on the AuthenticationController to remove the DuplicateKeyException on get type methods and added null checks on AuthenticationSerivce

**Pull request merged**